packages/eslint-plugin/src/rules/no-base-to-string.ts

```diff
@@ -82,6 +89,50 @@ export default createRule<Options, MessageIds>({
       });
     }
 
+    function checkExpressionForArrayJoin(
+      node: TSESTree.Node,
+      type: ts.Type,
+    ): void {
+      const certainty = collectJoinCertainty(type);
+
+      if (certainty === Usefulness.Always) {
+        return;
+      }
+
+      context.report({
+        node,
+        messageId: 'baseArrayJoin',
+        data: {
+          name: context.sourceCode.getText(node),
+          certainty,
+        },
+      });
+    }
+
+    function collectJoinCertainty(type: ts.Type): Usefulness {
+      const typeParts = tsutils.typeParts(type);
+
+      const isArrayOrTuple = typeParts.every(
```

Comment:
@kirkwaiblinger 
> Here's what I hacked out... There's room for refinement in terms of verbosity, duplication, and short-circuiting, but I think that it handles correctness.

This sounds good, but I think it would be better to handle the intersection case differently, considering the following case.

```ts
class Foo {}
declare const tuple: [string] & [Foo];
tuple.join(''); // should be passed without error
```

```ts
     //...
      if (checker.isTupleType(type)) {
        const typeArgs = checker.getTypeArguments(type);
        const certainties = typeArgs.map(t => collectToStringCertainty(t));
        if (certainties.some(certainty => certainty === Usefulness.Never)) {
          return Usefulness.Never;
        }

        if (certainties.some(certainty => certainty === Usefulness.Sometimes)) {
          return Usefulness.Sometimes;
        }

        return Usefulness.Always;
      }
      
       //...
```